### ISSUE TITLE
The matching algorithm in findValues.ts calculates wrong scoring.

### ISSUE THREAD

**Issue by kichinsky** [microsoft/botbuilder-js#742]
## Versions
BF 4.2, Typescript

## Describe the bug
The algorithm implemented in the [findValues.ts](https://github.com/Microsoft/botbuilder-js/blob/d4255c382189476dff5cc0879ec8d2f29b1162ce/libraries/botbuilder-dialogs/src/choices/findValues.ts) ignores any leading words when calculating the distance, that might be not expected behaviour. It computes the distance only between tokens in the choice option and ignores the structure of utterance.

## To Reproduce
1. Implement the ChoicePrompt with one of the options equal to "Navigate". 
2. Type something like "bla navigate lala".
3. Check step.result.score in the next step of the WaterfallPrompt.

## Expected behavior
The score in step.result after the ChoicePrompt executed should not be equal to 1. So the leading "bla" word should be taken into account when calculating the score.

Similar, if one of the choice options is "Band Search" and the utterance it "blala band bla search" the score should be less, than 2/3 (as it is now) because of the leading "blala" word.

[bug]

**Comment by ceciliaavila** [comment]
Hi @kichinsky, @Stevenic,
We reviewed this issue and found that the **score** in step.result is not the position where the matching word is. It's a number that represents the quality of the match. It's calculated as:
![image](https://user-images.githubusercontent.com/44245136/54848546-51b8ad00-4cc0-11e9-934a-17cf8937e494.png)

This field is relevant when _allowPartialMatches_ option is set to true, where, in case of a partial match, the score would be less than 1.
For example, if the Option is "Adaptive Card" and the user types  "adaptive",  the score of the result would be 0.5
![image](https://user-images.githubusercontent.com/44245136/54848580-6b59f480-4cc0-11e9-9945-8cc604d29e94.png)

If you agree, would you please close this issue?

Thanks!

**Comment by kichinsky** [comment]
Hi @ceciliaavila, I didn't mean that the score is the position of the matching word. From your example I see that your calculation formula comes from the idea of partial matching, when you give user some "multi word options" and they are able to choose by typing just some of the "word"s.

My test use case was quite opposite: user actually types more words, than are expected in the choicePrompt, or something totally different by meaning but still containing some of the words. 

E.g.: one of the expected options is "Navigate", and user types "I wanna navigate". I wouldn't expect that the score is "1" in that case (that I interpret as perfect match).

**Comment by ceciliaavila** [comment]
Hi @kichinsky, @Stevenic,
We did some more tests and, as you mentioned, the score is the same (1) if the Choice's value matches complete or partially with the utterance from the user.
E.g.:

| Choice value | User's utterance  | Score |
| -------------- | ------------------ | ------ |
| Navigate |navigate | 1
| Navigate | web navigate | 1
| Web Navigate | navigate | 0 (*) 

(*) _AllowPartialMatches = false (default). If this option is set to true, then the score would be 0.5_

@tellarin, can you please tell us if this is the expected behavior or is it a bug that needs to be fixed? (This is implemented like this in DotNet as well).

Thanks!

**Comment by Stevenic** [comment]
This looks to be a bug. We should try to fix in 4.4 if we have the cycles.

**Comment by cleemullins** [comment]
@stevenic Any update on this?

**Comment by gabog** [comment]
@Stevenic, did you have a chance to take a look at this one?

**Comment by Stevenic** [comment]
Actually... Digging back into this code the behavior above is currently "By Design".  The score returned is our confidence that we found a given choice within a utterance.  So if we find all of the terms for a given choice our confidence is 1.0 that choice exists within the utterance.  To understand why this works the way it works lets look at an example involving multiple choices. Lets say our choices are:

```
['green peppers', 'red peppers', 'onions', 'tomatoes', 'lettuce', 'cheese']
```

With an utterance of "I would like lettuce, tomatoes, and peppers" we would get the following results returned:


Choice value | Score
-- | --
red peppers | 0.5
green peppers | 0.5
tomatoes | 1.0
lettuce | 1.0

This basically tells us that we found "tomatoes" and "lettuce" with high confidence but "red peppers" and "green peppers" were ambiguous. This could lead to us sending a clarifying question of "did you mean green peppers or red peppers".
